src/registry.rs

```diff
@@ -152,109 +180,150 @@ pub enum BindError {
 #[derive(Debug)]
 pub struct RegistryState {
     registry: wl_registry::WlRegistry,
-    cached_globals: HashMap<u32, CachedGlobal>,
+    globals: Vec<Global>,
+    ready: bool,
+}
+
+#[derive(Debug)]
+struct Global {
+    interface: String,
+    version: u32,
+    name: u32,
 }
 
 impl RegistryState {
     /// Creates a new registry handle.
     ///
     /// This type may be used to bind globals as they are advertised.
-    pub fn new(registry: wl_registry::WlRegistry) -> RegistryState {
-        RegistryState { registry, cached_globals: HashMap::new() }
+    pub fn new<D>(conn: &Connection, qh: &QueueHandle<D>) -> Self
+    where
+        D: Dispatch<wl_registry::WlRegistry, ()>
+            + Dispatch<wl_callback::WlCallback, RegistryReady>
+            + ProvidesRegistryState
+            + 'static,
+    {
+        let display = conn.display();
+        let registry = display.get_registry(qh, ()).unwrap();
+        display.sync(qh, RegistryReady).unwrap();
+        RegistryState { registry, globals: Vec::new(), ready: false }
     }
 
-    /// Binds a global, returning a new object associated with the global.
+    /// Uses an existing WlRegistry for handling registry state.
     ///
-    /// This function may be used for any global, but should be avoided if the global being bound may be used
-    /// by multiple modules of smithay's client toolkit. If multiple modules need a global, use
-    /// [`RegistryState::bind_cached`] instead.
+    /// Note: prefer using [Self::new] unless you need access to the registry for other reasons.
     ///
-    /// A protocol error will be risen if the global has not yet been advertised.
-    pub fn bind_once<I, D, U>(
-        &mut self,
+    /// You will need to ensure the RegistryReady signal is sent to this object after initial
+    /// enumeration of the registry is complete.
+    pub fn from_registry(registry: wl_registry::WlRegistry) -> Self {
+        RegistryState { registry, globals: Vec::new(), ready: false }
+    }
+
+    pub fn registry(&self) -> &wl_registry::WlRegistry {
+        &self.registry
+    }
+
+    /// Returns true if the registry has completed the initial enumeration of globals and is ready
+    /// to serve bind requests.
+    pub fn ready(&self) -> bool {
+        self.ready
+    }
+
+    /// Binds a global, returning a new object associated with the global.
+    ///
+    /// This should not be used to bind globals that have multiple instances such as `wl_output`;
+    /// use [Self::bind_all] instead.
+    pub fn bind_one<I, D, U>(
+        &self,
         qh: &QueueHandle<D>,
-        name: u32,
-        version: u32,
+        version: std::ops::RangeInclusive<u32>,
```

Comment:
Should we validate the range is valid (an assert)?

You could write `4..=1` and rust would not complain but it would bind the wrong version.

Reply:
I don't see that as a big issue - it's much less likely than just passing in nonsense as the version range endpoints, which we also don't detect.